src/uuid.ts

```diff
@@ -1,57 +1,211 @@
-/**
- * UUID regular expression pattern copied from `uuid` npm module.
- * @see https://github.com/uuidjs/uuid/blob/master/src/regex.js
- */
-const UUID_RX = /^(?:[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}|00000000-0000-0000-0000-000000000000)$/i;
+import { Buffer } from 'buffer';
+import { ensureBuffer } from './ensure_buffer';
+import { Binary, BinaryExtended, BinaryExtendedLegacy } from './binary';
+import { bufferToUuidHexString, uuidHexStringToBuffer, uuidValidateString } from './uuid_utils';
+import type { EJSONOptions } from './extended_json';
+import { randomBytes } from './parser/utils';
 
 /** @public */
-export interface UUIDExtended {
+export type UUIDExtended = {
   $uuid: string;
-}
+};
+
+const BYTE_LENGTH = 16;
+
+const kId = Symbol('id');
 
 /**
- * Parser function copied from `uuid` npm module.
- * @see https://github.com/uuidjs/uuid/blob/master/src/parse.js
- * @internal
+ * A class representation of the BSON UUID type.
+ * @public
  */
-export function parseUUID(uuid: string): Uint8Array {
-  if (typeof uuid !== 'string') {
-    throw new TypeError('Invalid type for UUID, expected string but got ' + typeof uuid);
+export class UUID {
+  _bsontype = 'UUID' as const;
```

Comment:
I see. I've removed all tests assuming a `UUID` can be serialized/deserialized, but kept 'uuid' out as a dependency, as it forced a certain string representation of uuids (including dashes) and duplicating a lot of functionality already present here.

Is the presence of a `_bsontype` property on an object seen as an indication it is directly serializable/deserializable?
I'm asking as we currently, in Realm, checks this property to detect if the type originates from this library, and which type. I guess my question boils down to: should this property be removed from the `UUID` class?

Reply:
I think its fine to leave it on the class for that purpose, even though it isn't directly used in the serializer, it would be good to leave a comment on the class that it isn't meant to be used for direct serialization.